src/type-exercises.ts

```diff
@@ -0,0 +1,249 @@
+ * This means the properties of the resulting type cannot be reassigned.
+ *
+ * @example
+ * interface Todo {
+ *   title: string;
+ *   description: string;
+ * }
+ *
+ * const todo: MyReadonly<Todo> = {
+ *   title: "Hey",
+ *   description: "foobar"
+ * };
+ *
+ * todo.title = "Hello";       // Error: cannot reassign a readonly property
+ * todo.description = "barFoo"; // Error: cannot reassign a readonly property
+ */
+
+// Add here your solution
+type MyReadonly<T> = {
+  readonly [P in keyof T]: T[P]
+}
+
+type ReadOnlyClient = MyReadonly<Client>
+const readClient = client as ReadOnlyClient
+
+console.log(readClient)
+
+//Now we can't modify any property of readClient
+//readClient.active = false
+
+// Add here your example
+
+/**
+ * Exercise #4: Recreate the built-in `ReturnType<T>` utility type without using it.
+ *
+ * @description
+ * The `MyReturnType<T>` utility type extracts the return type of a function type `T`.
+ *
+ * @example
+ * const fn = (v: boolean) => {
+ *   if (v) {
+ *     return 1;
+ *   } else {
+ *     return 2;
+ *   }
+ * };
+ *
+ * type a = MyReturnType<typeof fn>; // expected to be "1 | 2"
+ */
+
+// Add here your solution
+type MyReturnType<F> = F extends (...args: any) => infer R ? R : any
+const readOrEditClient = (operation: 'read' | 'edit') => {
+  if (operation === 'edit') {
+    const verifiedClient: VerifiedClient = client as VerifiedClient
+    return verifiedClient
+  } else {
+    const unverifiedClient: UnverifiedClient = client as UnverifiedClient
+    return unverifiedClient
+  }
+}
+
+type functionType = MyReturnType<typeof readOrEditClient>
+
+// Add here your example
+
+/**
+ * Exercise #5: Extract the type inside a wrapped type like `Promise`.
+ *
+ * @description
+ * Implement a utility type `MyAwaited<T>` that retrieves the type wrapped in a `Promise` or similar structure.
+ *
+ * If `T` is `Promise<ExampleType>`, the resulting type should be `ExampleType`.
+ *
+ * @example
+ * type ExampleType = Promise<string>;
+ *
+ * type Result = MyAwaited<ExampleType>; // expected to be "string"
+ */
+
+// Add here your solution
+type MyAwaited<T extends Promise<any>> = T extends Promise<infer U> ? U : never
+
+// Add here your example
+type MyAwaitedClient = MyAwaited<Promise<Client>>
+
+/**
+ * Exercise 6: Create a utility type `RequiredByKeys<T, K>` that makes specific keys of `T` required.
+ *
+ * @description
+ * The type takes two arguments:
+ * - `T`: The object type.
+ * - `K`: A union of keys in `T` that should be made required.
+ *
+ * If `K` is not provided, the utility should behave like the built-in `Required<T>` type, making all properties required.
+ *
+ * @example
+ * interface User {
+ *   name?: string;
+ *   age?: number;
+ *   address?: string;
+ * }
+ *
+ * type UserRequiredName = RequiredByKeys<User, 'name'>;
+ * // expected to be: { name: string; age?: number; address?: string }
+ */
+
+type RequiredByKeys<T, K extends keyof T> = Omit<T, K> & Required<Pick<T, K>>
```

Comment:
Good solution even though, the type inferred is not going to be simplified by the type and it is going to be show as the merge of the type.

Reply:
I've added some extra logic to the type in order to flatten its result. https://github.com/ravnhq/TS-Nederdy-Challenges/pull/52/commits/faf7bb5e960ac130c0c24f562b08cabae9d48f90